### PR TITLE
Implement paint and accessibility passes

### PR DESCRIPTION
**Pull request by PoignardAzur**

### CODE REVIEW

**Comment by PoignardAzur** [comment]
(This is very much a draft, more so than usual. Accessibility isn't ported yet, and there's a bunch of changes I deliberately delayed to avoid creating huge diffs.)

**Comment by PoignardAzur** [comment]
I think this is ready for review now. I left some things for later (notably removing `WidgetPod::paint` and `WidgetPod::accessibility` and using a more general shape for the clip path) because the diff is large enough as it is.

I don't intend to add a default implementation for paint and accessibility, because I think widgets that don't need to do anything with them will be rare (even eg Flex will get properties like a background color); for accessibility in particular, I think we should encourage people to look at accessibility tree values before decide they don't need any, and having a default impl goes against that.

File: masonry/src/contexts.rs
Comment (jaredoconnell) [inline masonry/src/contexts.rs:893]: I've seen this repeated at least three times. It may make sense to create a function just for marking the state change for needing paint and accessibility.

**Review by jaredoconnell — COMMENTED** [review]
Looks good. I just have one minor comment.

File: masonry/src/passes/event.rs
Comment (waywardmonkeys) [inline masonry/src/passes/event.rs:168]: (Had commented directly on the commit, but then that doesn't show up here ...)

I use this in my own event loop integration. I need some way to pump an access event into the system ...

Comment (waywardmonkeys) [reply masonry/src/passes/event.rs:168]: My mistake. I'm calling this on `RenderRoot`, not this one here.

File: masonry/src/contexts.rs
Comment (PoignardAzur) [reply masonry/src/contexts.rs:893]: I'd like to keep these kinds of refactor for last, especially if we're only factoring three lines or less. I think there's a benefit to being able to see directly what the code does instead of jumping to another function.

**Comment by DJMcNab** [comment]
First thought when reviewing. The only thing which renders in `mason` is the textbox, which I don't think is the intention

**Comment by PoignardAzur** [comment]
> The only thing which renders in mason is the textbox, which I don't think is the intention

This is also the case on main. Bisecting to find the culprit...

**Review by DJMcNab — APPROVED** [review]
This is looking really good! I don't have any blocking concerns, but there are a few bits of clarification I'd like to see.

File: masonry/src/passes/accessibility.rs
Comment (PoignardAzur) [reply masonry/src/passes/accessibility.rs:68]: I assume that by "these methods" you mean "the methods that gave me problems in the layout PR"?

To be honest, I haven't decided on (or documented) a principled solution for that class of problem. I'm mostly deferring it until all passes are implemented.

But yeah, your reasoning makes sense.

File: masonry/src/widget/widget_arena.rs
Comment (DJMcNab) [inline masonry/src/widget/widget_arena.rs:16]: Do these ever get cleaned up?

File: masonry/src/widget/widget_state.rs
Comment (PoignardAzur) [reply masonry/src/widget/widget_state.rs:68]: I'd rather support arbitrary shapes. Or at the very least, rectangles with rounded corners. The reason I'm doing rectangles only is that Kurbo doesn't provide you with a type to store an arbitrary shape.

File: masonry/src/passes/paint.rs
Comment (PoignardAzur) [reply masonry/src/passes/paint.rs:35]: I think it was used in Druid for z-order stuff? I guess it's obsolete now.

File: masonry/src/passes/paint.rs
Comment (PoignardAzur) [reply masonry/src/passes/paint.rs:49]: ... Oh, good point.

File: masonry/src/passes/paint.rs
Comment (PoignardAzur) [reply masonry/src/passes/paint.rs:78]: There's also the fact that, if we adopt compositor layers, then you'll want to display *some* offscreen stuff so you don't get caught flatfooted when the user scrolls.

File: masonry/src/widget/widget_arena.rs
Comment (PoignardAzur) [reply masonry/src/widget/widget_arena.rs:16]: Good catch.

File: masonry/src/widget/widget_arena.rs
Comment (PoignardAzur) [reply masonry/src/widget/widget_arena.rs]: I removed them.

File: masonry/src/passes/paint.rs
Comment (PoignardAzur) [reply masonry/src/passes/paint.rs]: Removed it.

File: masonry/src/passes/paint.rs
Comment (PoignardAzur) [reply masonry/src/passes/paint.rs:76]: I've added a comment here.

**Comment by PoignardAzur** [comment]
I think I've addressed all review comments.

This is a big one, so I'll wait until tomorrow to merge it, in case there's any concerns left.

File: masonry/src/passes/accessibility.rs
Comment (DJMcNab) [reply masonry/src/passes/accessibility.rs:68]: In this case, I meant `request_accessibility`. I hadn't checked if `get_raw_mut` were available in `AccessCtx`

File: masonry/src/widget/widget_state.rs
Comment (DJMcNab) [reply masonry/src/widget/widget_state.rs:68]: Right - I guess a solution for rounded rectangles and similar is to draw the parent surface "above" the child surfaces, and have a transparent cutout for the composited view. If compositors can implement that performantly, that would work well

My mistake, I think arbitrary paths are probably fine then, although it probably requires some careful use of blend modes to punch through a transparent region.

**Pull request merged**
